difficulties building with 32-bit mingw-i686

I've been trying to get the OpenCL ICD loader building and working properly with 32-bit mingw-i686 as part of https://github.com/KhronosGroup/OpenCL-SDK/issues/58 but I seem to have hit a bit of a dead end, so I'm at least documenting what I've tried in case it is helpful.

I'm cross-compiling on a Linux host (Ubuntu 20.04).  I have the mingw-w64 packages installed.

```sh
$ i686-w64-mingw32-g++ --version
i686-w64-mingw32-g++ (GCC) 9.3-posix 20200320
```

I'm trying to build my [SimpleOpenCLSamples](https://github.com/bashbaug/SimpleOpenCLSamples), which builds with the upstream OpenCL-Headers and OpenCL-ICD-Loader.  The specific commits I am using are:

* SimpleOpenCLSamples - https://github.com/bashbaug/SimpleOpenCLSamples/commit/42ca56455295414bb926f8aec9a2d32a8fe0af4f (latest in the main branch)
* OpenCL-Headers - https://github.com/KhronosGroup/OpenCL-Headers/commit/b590a6bfe034ea3a418b7b523e3490956bcb367a (latest in the main branch)
* OpenCL-ICD-Loader - https://github.com/bashbaug/OpenCL-ICD-Loader/commit/6c51a7912d2a57754f6e8bf1bb4159458768f2a1 (latest in a branch with a few mingw fixes/improvements)

With these commits I am able to build an OpenCL.dll and a libOpenCL.dll.a import library, but I get an error when I link even the simplest sample with these libraries:

```sh
$ make clean enumopencl
[  8%] Building C object external/opencl-icd-loader/CMakeFiles/OpenCL.dir/loader/icd.c.obj
...
[ 83%] Linking C shared library OpenCL.dll
[ 83%] Built target OpenCL
[ 91%] Building CXX object samples/00_enumopencl/CMakeFiles/enumopencl.dir/main.cpp.obj
[100%] Linking CXX executable enumopencl.exe
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x148): undefined reference to `clGetPlatformInfo@20'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x1ad): undefined reference to `clGetPlatformInfo@20'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x350): undefined reference to `clGetDeviceInfo@20'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x3b5): undefined reference to `clGetDeviceInfo@20'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x5d2): undefined reference to `clGetDeviceInfo@20'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0x9af): undefined reference to `clGetPlatformIDs@12'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0xa17): undefined reference to `clGetPlatformIDs@12'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0xab2): undefined reference to `clGetDeviceIDs@24'
/usr/bin/i686-w64-mingw32-ld: CMakeFiles/enumopencl.dir/objects.a(main.cpp.obj):main.cpp:(.text+0xb3e): undefined reference to `clGetDeviceIDs@24'
collect2: error: ld returned 1 exit status
make[3]: *** [samples/00_enumopencl/CMakeFiles/enumopencl.dir/build.make:101: samples/00_enumopencl/enumopencl.exe] Error 1
make[2]: *** [CMakeFiles/Makefile2:1326: samples/00_enumopencl/CMakeFiles/enumopencl.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1333: samples/00_enumopencl/CMakeFiles/enumopencl.dir/rule] Error 2
make: *** [Makefile:374: enumopencl] Error 2
```

The problem appears to be due to usage of the `stdcall` calling convention for 32-bit Windows, though I'm not 100% sure where things are going wrong.  Here is my current understanding:

1. The application itself is looking for a decorated function name (with the `@n` suffix for stdcall) and is not finding it, hence the linker failure.

2. The import library libOpenCL.dll.a does NOT decorate the functions it exports:

```sh
$ winedump -j export external/opencl-icd-loader/libOpenCL.dll.a 
Contents of external/opencl-icd-loader/libOpenCL.dll.a: 81250 bytes

248 public symbols
    1bd4 _OpenCL_dll_iname
    1d0e __head_OpenCL_dll
    1e8a _clWaitForEvents
    1e8a __imp__clWaitForEvents
    20d2 _clUnloadPlatformCompiler
    20d2 __imp__clUnloadPlatformCompiler
...
```

3. The DLL itself also does NOT decorate the functions it exports:

```sh
$ winedump -j export external/opencl-icd-loader/OpenCL.dll 
Contents of external/opencl-icd-loader/OpenCL.dll: 419689 bytes

Exports table:

  Name:            OpenCL.dll
  Characteristics: 00000000
  TimeDateStamp:   6344ADAB Mon Oct 10 16:41:31 2022
  Version:         0.00
  Ordinal base:    1
  # of functions:  123
  # of Names:      123
Addresses of functions: 00015028
Addresses of name ordinals: 00015400
Addresses of names: 00015214

  Entry Pt  Ordn  Name
  00003EE2     1 clBuildProgram
  000085AC     2 clCloneKernel
  00006ED5     3 clCompileProgram
...
```

(2) and (3) seem to be consistent with the OpenCL.def in this repo, which also does not decorate the functions it exports:

```
$ more ../external/opencl-icd-loader/loader/windows/OpenCL.def 
; copyright text removed here

EXPORTS

; comment removed here

; OpenCL 1.0 API
clBuildProgram
clCreateBuffer
clCreateCommandQueue
clCreateContext
...
```

I suppose this could still be the problem though?

Some things I tried:

* Adding the linker option `--add-stdcall-alias` (didn't seem to have an effect).
* Adding the linker option `--disable-stdcall-fixup` (couldn't link OpenCL.dll).
* Various combinations with the linker option `--kill-at` (didn't seem to have an affect).
* Changing OpenCL.def to explicitly add an alias to the decorated function (didn't seem to have an affect):

    ```
    clGetPlatformIDs = clGetPlatformIDs@12
    ```

* Changing OpenCL.def to only include the decorated function:

    ```
    clGetPlatformIDs@12
    ```

    I was able to link an executable with this change.  Not sure how viable of a solution this is, though...

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -80,13 +80,20 @@ if (WIN32)
         loader/windows/icd_windows_hkr.h
         loader/windows/icd_windows_apppackage.c
         loader/windows/icd_windows_apppackage.h
-        loader/windows/OpenCL.def
         loader/windows/OpenCL.rc)
     # Only add the DXSDK include directory if the environment variable is
     # defined.  Since the DXSDK has merged into the Windows SDK, this is
     # only required in rare cases.
     if (DEFINED ENV{DXSDK_DIR} AND NOT (MINGW OR MSYS OR CYGWIN))
         include_directories ($ENV{DXSDK_DIR}/Include)
+    endif ()
+
+    # For mingw-i686 builds only we need a special .def file with stdcall
+    # exports.  In all other cases we can use a standard .def file.
+    if (("${CMAKE_SIZEOF_VOID_P}" EQUAL "4") AND (MINGW OR MSYS OR CYGWIN))
+        list (APPEND OPENCL_ICD_LOADER_SOURCES loader/windows/OpenCL-mingw-i686.def)
+    else ()
+        list (APPEND OPENCL_ICD_LOADER_SOURCES loader/windows/OpenCL.def)
     endif ()
 else ()
     list (APPEND OPENCL_ICD_LOADER_SOURCES
@@ -115,7 +122,7 @@ if (WIN32)
     # Generate a DLL without a "lib" prefix for mingw.
     if (MINGW OR MSYS OR CYGWIN)
         set_target_properties(OpenCL PROPERTIES PREFIX "")
-        set_target_properties(OpenCL PROPERTIES LINK_FLAGS "-Wl,-enable-stdcall-fixup")
+        set_target_properties(OpenCL PROPERTIES LINK_FLAGS "-Wl,-disable-stdcall-fixup")
     endif()
 else()
     target_link_libraries (OpenCL PRIVATE ${CMAKE_THREAD_LIBS_INIT})

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -111,6 +111,12 @@ if (WIN32)
     target_link_libraries (OpenCL PRIVATE cfgmgr32.lib runtimeobject.lib)
 
     option (OPENCL_ICD_LOADER_DISABLE_OPENCLON12 "Disable support for OpenCLOn12. Support for OpenCLOn12 should only be disabled when building an import lib to link with, and must be enabled when building an ICD loader for distribution!" OFF)
+
+    # Generate a DLL without a "lib" prefix for mingw.
+    if (MINGW OR MSYS OR CYGWIN)
+        set_target_properties(OpenCL PROPERTIES PREFIX "")
+        set_target_properties(OpenCL PROPERTIES LINK_FLAGS "-Wl,-enable-stdcall-fixup")
+    endif()
 else()
     target_link_libraries (OpenCL PRIVATE ${CMAKE_THREAD_LIBS_INIT})
     if (NOT APPLE)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -90,7 +90,7 @@ if (WIN32)
 
     # For mingw-i686 builds only we need a special .def file with stdcall
     # exports.  In all other cases we can use a standard .def file.
-    if (("${CMAKE_SIZEOF_VOID_P}" EQUAL "4") AND (MINGW OR MSYS OR CYGWIN))
+    if ((CMAKE_SIZEOF_VOID_P EQUAL 4) AND (MINGW OR MSYS OR CYGWIN))
         list (APPEND OPENCL_ICD_LOADER_SOURCES loader/windows/OpenCL-mingw-i686.def)
     else ()
         list (APPEND OPENCL_ICD_LOADER_SOURCES loader/windows/OpenCL.def)

--- a/loader/icd_platform.h
+++ b/loader/icd_platform.h
@@ -40,4 +40,11 @@
 #error Unknown OS!
 #endif
 
+#ifdef __MINGW32__
+#if !defined(_WIN32_WINNT) || (_WIN32_WINNT < 0x0600)
+#undef _WIN32_WINNT
+#define _WIN32_WINNT 0x0600
+#endif
+#endif // __MINGW32__
+
 #endif

--- a/loader/windows/OpenCL-mingw-i686.def
+++ b/loader/windows/OpenCL-mingw-i686.def
@@ -1,0 +1,162 @@
+;
+; Copyright (c) 2022 The Khronos Group Inc.
+;
+; Licensed under the Apache License, Version 2.0 (the "License");
+; you may not use this file except in compliance with the License.
+; You may obtain a copy of the License at
+;
+;     http://www.apache.org/licenses/LICENSE-2.0
+;
+; Unless required by applicable law or agreed to in writing, software
+; distributed under the License is distributed on an "AS IS" BASIS,
+; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+; See the License for the specific language governing permissions and
+; limitations under the License.
+;
+; OpenCL is a trademark of Apple Inc. used under license by Khronos.
+
+EXPORTS
+
+;
+; Note: This is a special .def file that should only be needed for i686
+; (32-bit) mingw builds.  In this case we need to export the stdcall-
+; decorated functions. In all other cases we can use the standard .def
+; file that does not have decorated functions.
+;
+
+; OpenCL 1.0 API
+clBuildProgram@24 == clBuildProgram
+clCreateBuffer@24 == clCreateBuffer
+clCreateCommandQueue@20 == clCreateCommandQueue
+clCreateContext@24 == clCreateContext
+clCreateContextFromType@24 == clCreateContextFromType
+clCreateFromGLBuffer@20 == clCreateFromGLBuffer
+clCreateFromGLRenderbuffer@20 == clCreateFromGLRenderbuffer
+clCreateFromGLTexture2D@28 == clCreateFromGLTexture2D
+clCreateFromGLTexture3D@28 == clCreateFromGLTexture3D
+clCreateImage2D@36 == clCreateImage2D
+clCreateImage3D@44 == clCreateImage3D
+clCreateKernel@12 == clCreateKernel
+clCreateKernelsInProgram@16 == clCreateKernelsInProgram
+clCreateProgramWithBinary@28 == clCreateProgramWithBinary
+clCreateProgramWithSource@20 == clCreateProgramWithSource
+clCreateSampler@20 == clCreateSampler
+clEnqueueAcquireGLObjects@24 == clEnqueueAcquireGLObjects
+clEnqueueBarrier@4 == clEnqueueBarrier
+clEnqueueCopyBuffer@36 == clEnqueueCopyBuffer
+clEnqueueCopyBufferToImage@36 == clEnqueueCopyBufferToImage
+clEnqueueCopyImage@36 == clEnqueueCopyImage
+clEnqueueCopyImageToBuffer@36 == clEnqueueCopyImageToBuffer
+clEnqueueMapBuffer@44 == clEnqueueMapBuffer
+clEnqueueMapImage@52 == clEnqueueMapImage
+clEnqueueMarker@8 == clEnqueueMarker
+clEnqueueNDRangeKernel@36 == clEnqueueNDRangeKernel
+clEnqueueNativeKernel@40 == clEnqueueNativeKernel
+clEnqueueReadBuffer@36 == clEnqueueReadBuffer
+clEnqueueReadImage@44 == clEnqueueReadImage
+clEnqueueReleaseGLObjects@24 == clEnqueueReleaseGLObjects
+clEnqueueTask@20 == clEnqueueTask
+clEnqueueUnmapMemObject@24 == clEnqueueUnmapMemObject
+clEnqueueWaitForEvents@12 == clEnqueueWaitForEvents
+clEnqueueWriteBuffer@36 == clEnqueueWriteBuffer
+clEnqueueWriteImage@44 == clEnqueueWriteImage
+clFinish@4 == clFinish
+clFlush@4 == clFlush
+clGetCommandQueueInfo@20 == clGetCommandQueueInfo
+clGetContextInfo@20 == clGetContextInfo
+clGetDeviceIDs@24 == clGetDeviceIDs
+clGetDeviceInfo@20 == clGetDeviceInfo
+clGetEventInfo@20 == clGetEventInfo
+clGetEventProfilingInfo@20 == clGetEventProfilingInfo
+clGetExtensionFunctionAddress@4 == clGetExtensionFunctionAddress
+clGetGLObjectInfo@12 == clGetGLObjectInfo
+clGetGLTextureInfo@20 == clGetGLTextureInfo
+clGetImageInfo@20 == clGetImageInfo
+clGetKernelInfo@20 == clGetKernelInfo
+clGetKernelWorkGroupInfo@24 == clGetKernelWorkGroupInfo
+clGetMemObjectInfo@20 == clGetMemObjectInfo
+clGetPlatformIDs@12 == clGetPlatformIDs
+clGetPlatformInfo@20 == clGetPlatformInfo
+clGetProgramBuildInfo@24 == clGetProgramBuildInfo
+clGetProgramInfo@20 == clGetProgramInfo
+clGetSamplerInfo@20 == clGetSamplerInfo
+clGetSupportedImageFormats@28 == clGetSupportedImageFormats
+clReleaseCommandQueue@4 == clReleaseCommandQueue
+clReleaseContext@4 == clReleaseContext
+clReleaseEvent@4 == clReleaseEvent
+clReleaseKernel@4 == clReleaseKernel
+clReleaseMemObject@4 == clReleaseMemObject
+clReleaseProgram@4 == clReleaseProgram
+clReleaseSampler@4 == clReleaseSampler
+clRetainCommandQueue@4 == clRetainCommandQueue
+clRetainContext@4 == clRetainContext
+clRetainEvent@4 == clRetainEvent
+clRetainKernel@4 == clRetainKernel
+clRetainMemObject@4 == clRetainMemObject
+clRetainProgram@4 == clRetainProgram
+clRetainSampler@4 == clRetainSampler
+clSetCommandQueueProperty@20 == clSetCommandQueueProperty
+clSetKernelArg@16 == clSetKernelArg
+clUnloadCompiler@0 == clUnloadCompiler
+clWaitForEvents@8 == clWaitForEvents
+
+; OpenCL 1.1 API
+clCreateSubBuffer@24 == clCreateSubBuffer
+clCreateUserEvent@8 == clCreateUserEvent
+clEnqueueCopyBufferRect@52 == clEnqueueCopyBufferRect
+clEnqueueReadBufferRect@56 == clEnqueueReadBufferRect
+clEnqueueWriteBufferRect@56 == clEnqueueWriteBufferRect
+clSetEventCallback@16 == clSetEventCallback
+clSetMemObjectDestructorCallback@12 == clSetMemObjectDestructorCallback
+clSetUserEventStatus@8 == clSetUserEventStatus
+
+; OpenCL 1.2 API
+clCompileProgram@36 == clCompileProgram
+clCreateFromGLTexture@28 == clCreateFromGLTexture
+clCreateImage@28 == clCreateImage
+clCreateProgramWithBuiltInKernels@20 == clCreateProgramWithBuiltInKernels
+clCreateSubDevices@20 == clCreateSubDevices
+clEnqueueBarrierWithWaitList@16 == clEnqueueBarrierWithWaitList
+clEnqueueFillBuffer@36 == clEnqueueFillBuffer
+clEnqueueFillImage@32 == clEnqueueFillImage
+clEnqueueMarkerWithWaitList@16 == clEnqueueMarkerWithWaitList
+clEnqueueMigrateMemObjects@32 == clEnqueueMigrateMemObjects
+clGetExtensionFunctionAddressForPlatform@8 == clGetExtensionFunctionAddressForPlatform
+clGetKernelArgInfo@24 == clGetKernelArgInfo
+clLinkProgram@36 == clLinkProgram
+clReleaseDevice@4 == clReleaseDevice
+clRetainDevice@4 == clRetainDevice
+clUnloadPlatformCompiler@4 == clUnloadPlatformCompiler
+
+; OpenCL 2.0 API
+clCreateCommandQueueWithProperties@16 == clCreateCommandQueueWithProperties
+clCreatePipe@28 == clCreatePipe
+clCreateSamplerWithProperties@12 == clCreateSamplerWithProperties
+clEnqueueSVMFree@32 == clEnqueueSVMFree
+clEnqueueSVMMap@36 == clEnqueueSVMMap
+clEnqueueSVMMemcpy@32 == clEnqueueSVMMemcpy
+clEnqueueSVMMemFill@32 == clEnqueueSVMMemFill
+clEnqueueSVMUnmap@20 == clEnqueueSVMUnmap
+clGetPipeInfo@20 == clGetPipeInfo
+clSetKernelArgSVMPointer@12 == clSetKernelArgSVMPointer
+clSetKernelExecInfo@16 == clSetKernelExecInfo
+clSVMAlloc@20 == clSVMAlloc
+clSVMFree@8 == clSVMFree
+
+; OpenCL 2.1 API
+clCloneKernel@8 == clCloneKernel
+clCreateProgramWithIL@16 == clCreateProgramWithIL
+clEnqueueSVMMigrateMem@36 == clEnqueueSVMMigrateMem
+clGetDeviceAndHostTimer@12 == clGetDeviceAndHostTimer
+clGetHostTimer@8 == clGetHostTimer
+clGetKernelSubGroupInfo@32 == clGetKernelSubGroupInfo
+clSetDefaultDeviceCommandQueue@12 == clSetDefaultDeviceCommandQueue
+
+; OpenCL 2.2 API
+clSetProgramReleaseCallback@12 == clSetProgramReleaseCallback
+clSetProgramSpecializationConstant@16 == clSetProgramSpecializationConstant
+
+; OpenCL 3.0 API
+clCreateBufferWithProperties@28 == clCreateBufferWithProperties
+clCreateImageWithProperties@32 == clCreateImageWithProperties
+clSetContextDestructorCallback@12 == clSetContextDestructorCallback

--- a/loader/windows/icd_windows_hkr.c
+++ b/loader/windows/icd_windows_hkr.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2019 The Khronos Group Inc.
+ * Copyright (c) 2017-2022 The Khronos Group Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -214,8 +214,12 @@ bool khrIcdOsVendorsEnumerateHKR(void)
     ULONG szBuffer = 0;
 
     OLECHAR display_adapter_guid_str[MAX_GUID_STRING_LEN];
+#if defined(CM_GETIDLIST_FILTER_CLASS) && defined(CM_GETIDLIST_FILTER_PRESENT)
     ULONG ulFlags = CM_GETIDLIST_FILTER_CLASS |
                     CM_GETIDLIST_FILTER_PRESENT;
+#else
+    ULONG ulFlags = 0x300;
+#endif
 
     iret = StringFromGUID2(
         &GUID_DEVCLASS_DISPLAY,
@@ -275,8 +279,6 @@ bool khrIcdOsVendorsEnumerateHKR(void)
 
     for (PWSTR deviceId = deviceIdList; *deviceId; deviceId += wcslen(deviceId) + 1)
     {
-        DEVPROPTYPE devpropType;
-
         KHR_ICD_WIDE_TRACE(L"Device ID: %ls\n", deviceId);
 
         ret = CM_Locate_DevNodeW(&devinst, deviceId, 0);
@@ -338,17 +340,22 @@ bool khrIcdOsVendorsEnumerateHKR(void)
                     KHR_ICD_WIDE_TRACE(L"    deviceInstanceID: %ls\n", deviceInstanceID);
                 }
 
-                ret = CM_Get_DevNode_PropertyW(
+                ret = CM_Get_DevNode_Registry_PropertyW(
                     devchild,
-                    &DEVPKEY_Device_ClassGuid,
-                    &devpropType,
+                    CM_DRP_CLASSGUID,
+                    NULL,
                     (PBYTE)&guid,
                     &szGuid,
                     0);
 
-                if (CR_SUCCESS != ret ||
-                    !IsEqualGUID(&OCL_GUID_DEVCLASS_SOFTWARECOMPONENT, &guid))
+                if (CR_SUCCESS != ret)
                 {
+                    KHR_ICD_TRACE("    CM_Get_DevNode_Registry_PropertyW returned 0x%"PRIxDW", skipping device...\n", ret);
+                    continue;
+                }
+                else if (!IsEqualGUID(&OCL_GUID_DEVCLASS_SOFTWARECOMPONENT, &guid))
+                {
+                    KHR_ICD_TRACE("    GUID does not match, skipping device...\n");
                     continue;
                 }
 

